devtools/x/src/lint/guppy.rs

```diff
@@ -384,3 +384,118 @@ impl<'cfg> PackageLinter for OverlayFeatures<'cfg> {
 fn feature_str(feature: Option<&str>) -> &str {
     feature.unwrap_or("[base]")
 }
+
+/// Ensure that all unpublished packages only use path dependencies for workspace dependencies
+#[derive(Debug)]
+pub struct UnpublishedPackagesOnlyUsePathDependencies;
+
+impl Linter for UnpublishedPackagesOnlyUsePathDependencies {
+    fn name(&self) -> &'static str {
+        "unpublished-packages-only-use-path-dependencies"
+    }
+}
+
+impl PackageLinter for UnpublishedPackagesOnlyUsePathDependencies {
+    fn run<'l>(
+        &self,
+        ctx: &PackageContext<'l>,
+        out: &mut LintFormatter<'l, '_>,
+    ) -> Result<RunStatus<'l>> {
+        let metadata = ctx.metadata();
+
+        // Skip all packages which aren't 'publish = false'
+        if !matches!(metadata.publish(), Some(&[])) {
+            return Ok(RunStatus::Executed);
+        }
+
+        let no_version_req = VersionReq::parse(">=0.0.0").expect(">=0.0.0 should be a valid req");
```

Comment:
You could avoid doing this multiple times by stashing the `no_version_req` in a field in this struct, and adding a constructor. Not a big deal at all though.